src/auxiliary/math.jl

```diff
@@ -412,8 +412,15 @@ Given ε = 1.0e-4, we use the following algorithm.
         c3 = convert(RealT, -1 / 21) * (2 * gamma * (gamma - 2) - 9) * c2
         return 0.5f0 * (x + y) * @evalpoly(f2, 1, c1, c2, c3)
     else
-        return (gamma - 1) / gamma * (y^gamma - x^gamma) /
-               (y^(gamma - 1) - x^(gamma - 1))
+        if isinteger(gamma)
+            yg = y^(gamma - 1)
+            xg = x^(gamma - 1)
+        else
+            yg = exp((gamma - 1) * log(y)) # equivalent to y^gamma but faster for non-integers
+            xg = exp((gamma - 1) * log(x)) # equivalent to x^gamma but faster for non-integers
+        end
+        return (gamma - 1) / gamma * (yg * y - xg * x) /
+               (yg - xg)
```

Comment:
```suggestion
        return (gamma - 1) / gamma * (yg * y - xg * x) / (yg - xg)
```

Does this formatting work now?

Reply:
And does it make a difference if you avoid an additional division by rewriting this part as something like

```suggestion
        return (gamma - 1) * (yg * y - xg * x) / (gamma * (yg - xg))
```

?

Reply:
It looks like there's not a big difference:
```

julia> @inline function stolarsky_mean_1(x::RealT, y::RealT, gamma::RealT) where {RealT <: Real}
           epsilon_f2 = convert(RealT, 1.0e-4)
           f2 = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y) # f2 = f^2
           if f2 < epsilon_f2
               # convenience coefficients
               c1 = convert(RealT, 1 / 3) * (gamma - 2)
               c2 = convert(RealT, -1 / 15) * (gamma + 1) * (gamma - 3) * c1
               c3 = convert(RealT, -1 / 21) * (2 * gamma * (gamma - 2) - 9) * c2
               return 0.5f0 * (x + y) * @evalpoly(f2, 1, c1, c2, c3)
           else
               if isinteger(gamma)
               yg = y^(gamma-1)
               xg = x^(gamma-1)
               else
               yg = exp((gamma-1)*log(y))
               xg = exp((gamma-1)*log(x))
               end
               return (gamma - 1)/gamma * (yg*y - xg*x) /  (yg - xg) 
           end
       end
stolarsky_mean_1 (generic function with 1 method)

julia> @inline function stolarsky_mean_2(x::RealT, y::RealT, gamma::RealT) where {RealT <: Real}
           epsilon_f2 = convert(RealT, 1.0e-4)
           f2 = (x * (x - 2 * y) + y * y) / (x * (x + 2 * y) + y * y) # f2 = f^2
           if f2 < epsilon_f2
               # convenience coefficients
               c1 = convert(RealT, 1 / 3) * (gamma - 2)
               c2 = convert(RealT, -1 / 15) * (gamma + 1) * (gamma - 3) * c1
               c3 = convert(RealT, -1 / 21) * (2 * gamma * (gamma - 2) - 9) * c2
               return 0.5f0 * (x + y) * @evalpoly(f2, 1, c1, c2, c3)
           else
               if isinteger(gamma)
                   yg = y^(gamma-1)
                   xg = x^(gamma-1)
                   else
               yg = exp((gamma-1)*log(y))
               xg = exp((gamma-1)*log(x))
                   end
               return (gamma - 1) * (yg*y - xg*x) / (gamma *  (yg - xg) ) 
           end
       end
stolarsky_mean_2 (generic function with 1 method)

julia> @benchmark value = stolarsky_mean_1($(Ref(300.1))[], $(Ref(410.7))[], $(Ref(2.0))[])
BenchmarkTools.Trial: 10000 samples with 999 evaluations per sample.
 Range (min … max):  8.965 ns … 101.965 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     9.815 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   9.794 ns ±   1.531 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

      ▁▂   ▃█▄        ▆▇    ▃    ▆▁    ▂▁                     ▁
  ▃▂▃▃██▆▅▅███▇▇▇▆▄▄▅▅██▇▄▆▇██▆▅▇██▄▄▅▆██▅▆██▆▅▄▅▆▇▇▅▅▇▄▄▃▅▄▆ █
  8.96 ns      Histogram: log(frequency) by time      11.4 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark value = stolarsky_mean_2($(Ref(300.1))[], $(Ref(410.7))[], $(Ref(2.0))[])
BenchmarkTools.Trial: 10000 samples with 999 evaluations per sample.
 Range (min … max):   9.183 ns … 19.411 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     10.492 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   10.481 ns ±  0.363 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                ▁▂▁▁▁▅▄▄  ▄▂ ▁▁█▆▇▁      ▂▁▁   ▁▂ ▁           ▂
  ▄▁▁▁▃▅▄▄▃██▆▅▆█████████▇██▇██████▇██▇▇▆███▆▇█████▇▆▆▆▇▇▇▇▇▇ █
  9.18 ns      Histogram: log(frequency) by time      11.8 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark value = stolarsky_mean_1($(Ref(300.1))[], $(Ref(410.7))[], $(Ref(1.4))[])
BenchmarkTools.Trial: 10000 samples with 997 evaluations per sample.
 Range (min … max):  19.662 ns … 45.032 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.761 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   19.815 ns ±  0.562 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

    ▃█▂                                                        
  ▂▄███▄▂▂▂▁▂▂▂▁▁▁▂▁▁▂▂▂▂▁▂▁▁▁▂▁▁▁▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▂ ▂
  19.7 ns         Histogram: frequency by time        21.4 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark value = stolarsky_mean_2($(Ref(300.1))[], $(Ref(410.7))[], $(Ref(1.4))[])
BenchmarkTools.Trial: 10000 samples with 997 evaluations per sample.
 Range (min … max):  19.621 ns … 27.873 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.987 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   20.087 ns ±  0.432 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

    ▁▄▆▇█████▇▅▂                    ▁     ▁▁▂▁▁               ▃
  ▃▆████████████▇▆▃▆▆▅▅▄▅▆▆▄▆▆▅▆▇▇█▇████████████▇▆▆▅▃▆▅▄▆▄▅▄▆ █
  19.6 ns      Histogram: log(frequency) by time      22.1 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark value = stolarsky_mean_1($(Ref(300.1))[], $(Ref(410.7))[], $(Ref(1.7))[])
BenchmarkTools.Trial: 10000 samples with 997 evaluations per sample.
 Range (min … max):  19.658 ns … 27.584 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.765 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   19.884 ns ±  0.437 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

  ▁▆█▇▄                                  ▁▁                   ▁
  ██████▃▅▅▃▄▅▅▄▅▄▄▅▃▅▄▄▅▅▄▅▅▅▆▅▇██▆▇▆▇▆▇███▇█▆▅▅▂▃▅▅▄▅▅▆▃▄▃▄ █
  19.7 ns      Histogram: log(frequency) by time      21.8 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark value = stolarsky_mean_2($(Ref(300.1))[], $(Ref(410.7))[], $(Ref(1.7))[])
BenchmarkTools.Trial: 10000 samples with 997 evaluations per sample.
 Range (min … max):  19.642 ns … 42.111 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     19.983 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   20.010 ns ±  0.352 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

          ▁▂▅▅▇█▇▇▅▂                                           
  ▂▂▂▃▃▅▅▇███████████▅▃▂▂▂▂▁▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂ ▃
  19.6 ns         Histogram: frequency by time        21.2 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

Reply:
I am mainly looking at the median results. There, the first option seems to be best all the time?

Reply:
I agree with you, I will compare also full runs, to see if there are any noticeable differences.